lib/msal-common/src/cache/CacheManager.ts

```diff
@@ -241,31 +241,117 @@ export abstract class CacheManager implements ICacheManager {
-                });
+            const validAccounts = this.buildTenantProfiles(
+                allAccounts,
+                accountFilter
+            );
             return validAccounts[0];
         } else {
             return null;
         }
     }
 
+    /**
+     * Matches filtered account entities with cached ID tokens that match the tenant profile-specific account filters
+     * and builds the account info objects from the matching ID token's claims
+     * @param cachedAccounts
+     * @param accountFilter
+     * @returns Array of AccountInfo objects that match account and tenant profile filters
+     */
+    private buildTenantProfiles(
+        cachedAccounts: AccountEntity[],
+        accountFilter?: AccountFilter
+    ): AccountInfo[] {
+        return cachedAccounts
+            .map((accountEntity) => {
+                return this.getAccountInfoForTenantProfile(
+                    accountEntity,
+                    accountFilter
+                );
+            })
+            .filter((accountInfo) => {
+                // Tenant Profiles with matching ID tokens in the cache will have ID token claims
+                return accountInfo.idTokenClaims;
+            });
+    }
+
+    private getAccountInfoForTenantProfile(
+        accountEntity: AccountEntity,
+        accountFilter?: AccountFilter
+    ): AccountInfo {
+        const tenantProfileFilter: TenantProfileFilter = {
+            localAccountId: accountFilter?.localAccountId,
+            loginHint: accountFilter?.loginHint,
+            name: accountFilter?.name,
+            sid: accountFilter?.sid,
+        };
+        return this.getAccountInfoFromEntity(
+            accountEntity,
+            accountFilter?.tenantId,
+            tenantProfileFilter
+        );
+    }
+
     private getAccountInfoFromEntity(
-        accountEntity: AccountEntity
+        accountEntity: AccountEntity,
+        targetTenantId?: string,
+        tenantProfileFilter?: TenantProfileFilter
     ): AccountInfo {
-        const accountInfo = accountEntity.getAccountInfo();
-        const idToken = this.getIdToken(accountInfo);
+        let accountInfo = accountEntity.getAccountInfo();
+        const idToken = this.getIdToken(accountInfo, undefined, targetTenantId);
+
         if (idToken) {
-            accountInfo.idToken = idToken.secret;
-            accountInfo.idTokenClaims = extractTokenClaims(
+            const idTokenClaims = extractTokenClaims(
                 idToken.secret,
                 this.cryptoImpl.base64Decode
             );
+            // Tenant Profile filtering
+            if (tenantProfileFilter) {
+                if (
+                    !!tenantProfileFilter.localAccountId &&
+                    !this.matchLocalAccountId(
+                        idTokenClaims,
+                        tenantProfileFilter.localAccountId
+                    )
+                ) {
+                    return accountInfo;
```

Comment:
We're returning if the filter _does not_ match? Should this return only if all filters match or no filters are defined?

Reply:
If this is by design - can you add some comments explaining what's happening here?

Reply:
Changed to use a boolean function to make it clearer that it's checking if the ID token claims don't match the filter.